WordPress/openverse
# Add DAG to decode and deduplicate image tags with escaped literal unicode sequences

## Fixes



Fixes to https://github.com/WordPress/openverse/issues/4452 by @krysal

## Description


Introduce a new temporary DAG, `decode_and_deduplicate_image_tags`, to find tags with escaped literal Unicode sequences, and to process them into actual unicode strings.

I've introduced a function using PL/Python3u `ov_unistr`. See the documentation string on the task that creates it for motivation and details. It essentially implements part of the fix from https://github.com/WordPress/openverse/pull/4143, just the part that works on escaped unicode sequences.

See the previous versions of this PR description (and the discussion below) for information on why we are not touching unescaped sequences. The short version is: it's entirely unsafe to guess what is unescaped unicode and what is just regular text.

## Testing Instructions

To test this locally:

1. Initialise the upstream DB with the new sample data, `./ov just down -v && ./ov just catalog/init`.
2. Run `./ov just catalog/pgcli` and execute `select jsonb_array_elements(tags) from image where identifier = 'aeba0547-61da-42ee-b561-27c8fc817d5a';` to observe the testing data. Note the following:
  - `muséo` already exists in processed unicode for two providers
  - The same string also exists as escaped unicode as well as unescaped unicode, for the flickr provider.
3. Now open Airflow (localhost:9090) and login with airflow/airflow as username and password. Enable the `batched_update` DAG and the `decode_and_deduplicate_image_tags` DAG. Run the latter and wait for it to complete.
4. Run the select from step 2 again, and confirm the following:
  - The unescaped unicode strings are unmodified and still present
  - `muséo` only exists once per provider
  - There are no escaped literal unicode sequences remaining in the tags
 

## Checklist


      PRs) or the media properties generator (`just catalog/generate-docs media-props`
      for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
